README.md
## About
`jmatch` is a library for doing pattern matching and value grouping in JavaScript with a convenient syntax. It's still very much in development 👷‍♂️.

CHANGELOG.md
## 0.2.0 (2020-04-18)

First minor update

#### :bug: Bug Fix

- `match`
  - [#5](https://github.com/rlnsy/jmatch/issues/5) Allow newlines in captured values


test/match.spec.ts
/**
 * Tests for the primary match function
 */

import { expect } from 'chai';
import 'mocha';
import { match, _ } from '../index';

describe("Match function", () => {

    it("Should correctly match basic tokens", () => {
        let ran = false;
        match("my name is Rowan and I'm 21 years old", [
            ["my name is ,1; and I'm ,2; years old",
                (name, age) => {
                    ran = true;
                    expect(age).to.equal("21");
                    expect(name).to.equal("Rowan");
                }]
        ]);
        if (!ran) {
            expect.fail("Did not run case code");
        }
    });

    it("Should correctly order values based on numeric tokens", () => {
        let ran = false;
        match("my name is Rowan and I'm 21 years old", [
            ["my name is ,2; and I'm ,1; years old",
                (age, name) => {
                    ran = true;
                    expect(age).to.equal("21");
                    expect(name).to.equal("Rowan");
                }]
        ]);
        if (!ran) {
            expect.fail("Did not run case code");
        }
    });

    it("Should match a default case id one is provided", () => {
        let ran = false;
        match("my name is Rowan and Im 21 years old", [
            ["my name is ,2; and I'm ,1; years old",
                (age, name) => { expect.fail("Ran wrong case code"); }],
            [_, () => {
                ran = true;
            }]
        ]);
        if (!ran) {
            expect.fail("Did not run case code");
        }
    });

    it("Should raise an Error if no match is hit", () => {
        try {
            match("my name is Rowan and Im 21 years old", [
                ["my name is ,2; and I'm ,1; years old", 
                    (age, name) => {
                        expect.fail("Ran case code");
                    }]
            ]);
            expect.fail("Did not raise an exception");
        } catch (err) {
            expect(err).to.be.an.instanceOf(Error);
        }
    });

    it("Should allow token-reserved characters to be escaped", () => {
        let ran = false;
        match("my name is Rowan and I'm 21 years old,22;", [
            ["my name is ,1; and I'm ,2; years old,22;",
                (name, age) => {
                    ran = true;
                    expect(age).to.equal("21");
                    expect(name).to.equal("Rowan");
                }]
        ]);
        if (!ran) {
            expect.fail("Did not run case code");
        }
    });

    it("Should be able to match multiline strings by default", () => {
        let ran = false;
        match("my name is\n Rowan and I'm 21 \nyears old,22;", [
            ["my name is\n ,1; and I'm ,2; \nyears old,22;",
                (name, age) => {
                    ran = true;
                    expect(age).to.equal("21");
                    expect(name).to.equal("Rowan");
                }]
        ]);
        if (!ran) {
            expect.fail("Did not run case code");
        }
    });

    it("Should not match multiline strings if newline not in pattern", () => {
        try {
            match("my name is\n Rowan and I'm 21 \nyears old,22;", [
                ["my name is ,1; and I'm ,2; years old,22;",
                    () => {
                        expect.fail("Ran case code")
                    }]
            ]);
        } catch (err) {
            expect(err).to.be.an.instanceOf(Error);
        }
    });

    it("Should allow multiline strings at the end of grouped values", () => {
        let ran = false;
        match("my name is Rowan\n and I'm 21 years old,22;", [
            ["my name is ,1; and I'm ,2; years old,22;",
                (name) => {
                    ran = true;
                    expect(name).to.equal("Rowan\n");
                }]
        ]);
        if (!ran) {
            expect.fail("Did not run case code");
        }
    });

    it("Should allow multiline strings in the middle of grouped values", () => {
        let ran = false;
        match("my name is Rowan\nLindsay and I'm 21 years old,22;", [
            ["my name is ,1; and I'm ,2; years old,22;",
                (name) => {
                    ran = true;
                    expect(name).to.equal("Rowan\nLindsay");
                }]
        ]);
        if (!ran) {
            expect.fail("Did not run case code");
        }
    });

    it("Should allow escaped quotes to be matched", () => {
        let ran = false;
        match("my name is \"Rowan\" and I'm 21 years old,22;", [
            ["my name is \",1;\" and I'm ,2; years old,22;",
                (name) => {
                    ran = true;
                    expect(name).to.equal("Rowan");
                }]
        ]);
        if (!ran) {
            expect.fail("Did not run case code");
        }
    });

});


src/match.ts
/**
 * Define types and functions for the match
 * module
 */

const TOKEN = /,[0-9]+;/g;
type PATTERN = string;
type VALUE = any;
type RHS = (...args: VALUE[]) => any;
type MATCH = [PATTERN, RHS];
type MATCH_STMNT = MATCH[];
type INPUT = string;
type ARGV = string | undefined

export const _ = "ELSE_CASE";

/**
 * Search for instances of value tokens in
 * the string s and replace them with regex
 * wildcard capture groups
 */
function detokenize(s: string): string {
    let result = s;
    while (result.search(TOKEN) > 0) {
        result = result.replace(TOKEN, '(.+)')
    }
    return result;
}

function preprocess(p: string) {
    return p.replace('\\,', ',').replace('\\;', ';');
}

/**
 * 
 * @param s tokens follow the regular expression above
 * and must be of the form ,<number>;
 * ',' and ';' are therefore reserved and must be double
 * escaped (e.q. \\;) to be recoqnized by the matcher 
 * @param m sequence of match cases; see types
 */
export function match(s: INPUT, m: MATCH_STMNT): any {
    if (!s || !m) {
        throw new Error("Missing one or more arguments to match");
    }
    let outcome = null;
    let i = 0;
    while (i < m.length && !outcome) {
        let p: PATTERN;
        let r: RHS;
        [p, r] = m[i];
        if (p == _) {
            outcome = () => r();
        } else {
            let argixs: number[] = [];
            let tokenMatch: RegExpExecArray | null;
            while ((tokenMatch = TOKEN.exec(p)) !== null) {
                const tokenLen = tokenMatch[0].length;
                const argi: number = Number(
                    tokenMatch[0].substr(1,tokenLen - 2));
                argixs.push(argi);
            }
            const mchStr: string = preprocess(detokenize(p));
            const inputMatch = new RegExp(mchStr, 's').exec(s);
            if (inputMatch) {
                let args: ARGV[] = argixs.map((i) => undefined)
                let vix = 0;
                while (vix < argixs.length) {
                    const v = inputMatch[vix + 1];
                    const argix = argixs[vix];
                    args[argix - 1] = v;
                    vix++;
                }
                outcome = () => r(...args);
            } else {
                i++; // check other matches
            }
        }
    }
    if (!outcome) {
        throw new Error(`No match for '${s}'`);
    } else {
        return outcome();
    }
}


scratch.ts
const TOKEN = /,[0-9]+;/g;
type PATTERN = string;
type VALUE = any;
type RHS = (...args: VALUE[]) => any;
type MATCH = [PATTERN, RHS];
type MATCH_STMNT = MATCH[];
type INPUT = string;
type ARGV = string | undefined

export const _ = "ELSE_CASE";

/**
 * Search for instances of value tokens in
 * the string s and replace them with regex
 * wildcard capture groups
 */
function detokenize(s: string): string {
    let result = s;
    while (result.search(TOKEN) > 0) {
        result = result.replace(TOKEN, '(.+)')
    }
    return result;
}

function preprocess(p: string) {
    return p.replace('\\,', ',').replace('\\;', ';');
}

/**
 * 
 * @param s tokens follow the regular expression above
 * and must be of the form ,<number>;
 * ',' and ';' are therefore reserved and must be double
 * escaped (e.q. \\;) to be recoqnized by the matcher 
 * @param m 
 */
export function match(s: INPUT, m: MATCH_STMNT): any {
    let outcome = null;
    let i = 0;
    while (i < m.length && !outcome) {
        let p: PATTERN;
        let r: RHS;
        [p, r] = m[i];
        if (p == _) {
            outcome = () => r();
        } else {
            let argixs: number[] = [];
            let tokenMatch: RegExpExecArray | null;
            while ((tokenMatch = TOKEN.exec(p)) !== null) {
                const tokenLen = tokenMatch[0].length;
                const argi: number = Number(
                    tokenMatch[0].substr(1,tokenLen - 2));
                argixs.push(argi);
            }
            const mchStr: string = preprocess(detokenize(p));
            console.log(mchStr);
            const inputMatch = new RegExp(mchStr).exec(s);
            if (inputMatch) {
                let args: ARGV[] = argixs.map((i) => undefined)
                let vix = 0;
                while (vix < argixs.length) {
                    const v = inputMatch[vix + 1];
                    const argix = argixs[vix];
                    args[argix - 1] = v;
                    vix++;
                }
                outcome = () => r(...args);
            } else {
                i++; // check other matches
            }
        }
    }
    if (!outcome) {
        throw new Error(`No match for '${s}'`);
    } else {
        return outcome();
    }
}

console.log("---Match case");

match("my name is Rowan and I'm 21 years old", [
    ["my name is ,2; and I'm ,1; years old", 
        (age, name) => { 
            console.log(name); 
            console.log(age); 
        }],
    [_, () => { console.log("default case"); }]
]);

console.log("---Default case");

match("my name is Rowan and Im 21 years old", [
    ["my name is ,2; and I'm ,1; years old", 
        (age, name) => { 
            console.log(name); 
            console.log(age); 
        }],
    [_, () => { console.log("default"); }]
]);

console.log("---No match case");

match("my name is Rowan and Im 21 years old", [
    ["my name is ,2; and I'm ,1; years old", 
        (age, name) => { 
            console.log(name); 
            console.log(age); 
        }]
]);

console.log("---Escape case");

match("my name is Rowan and I'm 21 years old,22;", [
    ["my name is ,2; and I'm ,1; years old\\,22\\;", 
        (age, name) => { 
            console.log(name); 
            console.log(age); 
        }]
]);
